source/common/router/req_header_formatter.h

```diff
@@ -0,0 +1,88 @@
+#pragma once
+
+#include <cstdint>
+#include <iostream>
+#include <string>
+#include <unordered_map>
+
+#include "common/common/logger.h"
+#include "common/http/access_log/access_log_formatter.h"
+#include "common/json/json_loader.h"
+
+namespace Envoy {
+namespace Router {
+
+/**
+ * Interface for all types of header formatters used for custom request headers.
+ */
+class HeaderFormatter {
+public:
+  virtual ~HeaderFormatter() {}
+
+  virtual const std::string
+  format(const Envoy::Http::AccessLog::RequestInfo& request_info) const PURE;
+};
+
+typedef std::shared_ptr<HeaderFormatter> HeaderFormatterPtr;
```

Comment:
can you try, map_.insert(std::make_pair(Http::LowerCaseString(header->getString("key")), std::move(header_formatter))); ?